### PR TITLE
#195 | topic preview implementation

### PR DESCRIPTION
**Pull request by adamdubiel**
Proper implementation of topic messages preview. Tested concurrently using `wrk` (10 conns, 2 threads), no ill side effects found.

Implementation is based on Guavas  `AtomicLongMap`, which limits the amount of messages saved per topic. Preview messages are then held inside single `ConcurrentLinkedDeque`, which makes it effective memory-wise. When persisting, queue is drained and snapshot with topic-oriented structure is created.

### CODE REVIEW

**Comment by tmmi** [comment]
👍

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/config/Configs.java [hunk 1]
```diff
@@ -92,8 +92,9 @@
     FRONTEND_SSL_TRUSTSTORE_PASSWORD("frontend.ssl.truststore.password", "password"),
     FRONTEND_SSL_TRUSTSTORE_FORMAT("frontend.ssl.truststore.format", "JKS"),
 
-    FRONTEND_MESSAGE_PREVIEW_ENABLED("frontend.message.preview.enabled", false),
-    FRONTEND_MESSAGE_PREVIEW_LOG_PERSIST_PERIOD_MS("frontend.message.preview.log.persist.period.ms", 30000),
+    FRONTEND_MESSAGE_PREVIEW_ENABLED("frontend.message.preview.enabled", true),
```
Comment (chemicL) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/config/Configs.java]: The docs say false, and it should be false.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/infrastructure/zookeeper/ZookeeperBasedRepository.java [hunk 2]
```diff
@@ -34,6 +36,17 @@ protected void ensureConnected() {
         }
     }
 
+    protected void ensurePathExists(String path) {
+        ensureConnected();
+        if (!pathExists(path)) {
+            try {
+                zookeeper.create().creatingParentsIfNeeded().forPath(path);
+            } catch (Exception e) {
```
Comment (chemicL) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/infrastructure/zookeeper/ZookeeperBasedRepository.java]: IOException is imported but generic Exception is caught here, which one do we want?

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/config/Configs.java (hunk 1 above)
Comment (adamdubiel) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/config/Configs.java]: ah, leftover after testing :)

File: docs/docs/configuration/topic-preview.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Topic preview
+
+See [Topic preview](/user/topic-preview) section for description of this feature.
+
+Option                                      | Description                                        | Default value
+------------------------------------------- | -------------------------------------------------- | -------------
+frontend.message.preview.enabled            | enable message preview                             | false
+frontend.message.preview.size               | maximum number of messages per topic               | 3
+frontend.message.preview.log.persist.period | how often to refresh previewed messages in seconds | 30
```
Comment (chemicL) [inline docs/docs/configuration/topic-preview.md]: We could add "seconds" suffix for ease of understanding without checking the docs.

**Comment by chemicL** [comment]
👍

File: docs/docs/configuration/topic-preview.md (hunk 3 above)
Comment (adamdubiel) [inline docs/docs/configuration/topic-preview.md]: done

**Pull request merged**